src/test/app/XChain_test.cpp

```diff
@@ -0,0 +1,4764 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/beast/unit_test/suite.hpp>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Issue.h>
+#include <ripple/protocol/SField.h>
+#include <ripple/protocol/STXChainAttestationBatch.h>
+#include <ripple/protocol/STXChainBridge.h>
+#include <ripple/protocol/Serializer.h>
+#include <ripple/protocol/TER.h>
+#include <ripple/protocol/XChainAttestations.h>
+
+#include <test/jtx.h>
+#include <test/jtx/Env.h>
+#include <test/jtx/attester.h>
+#include <test/jtx/multisign.h>
+#include <test/jtx/xchain_bridge.h>
+
+#include <functional>
+#include <limits>
+#include <optional>
+#include <random>
+#include <string>
+#include <tuple>
+#include <variant>
+#include <vector>
+
+#include <fstream>
+#include <iostream>
+
+namespace ripple::test {
+
+template <class T>
+struct xEnv : public jtx::XChainBridgeObjects
+{
+    jtx::Env env_;
+
+    xEnv(T& s, bool side = false)
+        : env_(s, jtx::envconfig(jtx::port_increment, side ? 3 : 0), features)
+    {
+        using namespace jtx;
+        STAmount xrp_funds{XRP(10000)};
+
+        if (!side)
+        {
+            env_.fund(xrp_funds, mcDoor, mcAlice, mcBob, mcCarol, mcGw);
```

Comment:
I've spent some time writing temporary unit tests as a way to study XChain.  The basics of `xEnv` have been really handy.  So thanks for doing that.

However, having the `xEnv` pre-fund magic accounts lead me to avoid using these magic account names all together.  I would much prefer if each unit test defined and funded only the accounts that it needs for the test.  I felt like these accounts (and particularly having them funded in a hidden fashion) were just land mines I had to learn to step around.

My 2 cents.

Reply:
I think we do need to refactor the tests, but I'd like to do that in a future PR. It's too much work to get that into this PR and the tests as written are Ok, even if not ideal.